### ISSUE TITLE
[SIP] patch for multiline headers

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2198]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)| @pitrou reported|
|-|-|
|Trac ID|trac#2198|
|Type|enhancement|
|Created|2006-10-27 17:10:38Z|
|Branch|https://github.com/twisted/twisted/tree/sip-multiline-2198|

Here is a patch for multiline headers.

For unit testing, add the following at the beginning of test_sip.py:
```
# multiline headers (example from RFC 3261)
response_multiline = """\
SIP/2.0 200 OK
Via: SIP/2.0/UDP server10.biloxi.com
    ;branch=z9hG4bKnashds8;received=192.0.2.3
Via: SIP/2.0/UDP bigbox3.site3.atlanta.com
    ;branch=z9hG4bK77ef4c2312983.1;received=192.0.2.2
Via: SIP/2.0/UDP pc33.atlanta.com
    ;branch=z9hG4bK776asdhds ;received=192.0.2.1
To: Bob <sip:bob@...>;tag=a6c85cf
From: Alice <sip:alice@...>;tag=1928301774
Call-ID: a84b4c76e66710@...
CSeq: 314159 INVITE
Contact: <sip:bob@192.0.2.4>
Content-Type: application/sdp
Content-Length: 0
\n""".replace("\n", "\r\n")
```

and then the following method in MessageParsingTestCase:

```
    def testMultiLine(self):
        l = self.l
        self.feedMessage(response_multiline)
        self.assertEquals(len(l), 1)
        m = l[0]
        self.assertEquals(m.headers['via'][0],
            "SIP/2.0/UDP server10.biloxi.com;branch=z9hG4bKnashds8;received=192.0.2.3")
        self.assertEquals(m.headers['via'][2],
            "SIP/2.0/UDP pc33.atlanta.com;branch=z9hG4bK776asdhds ;received=192.0.2.1")
```

Attachments:

* [patch-sip-multiline.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c07/c076cfc485c8ba2f77ebc74a81dcd94bc3eb540d/e576e15dfb1f673cef4f967933788db93702481f.txt) (2259 bytes) - added by antoine on 2006-10-27 17:11:32Z - patch to enable multiline headers for SIP
* [Twisted_2198.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c07/c076cfc485c8ba2f77ebc74a81dcd94bc3eb540d/b23a5c903c56218efb93bcf9f5a6785106f410f8.patch) (4031 bytes) - added by moijes12 on 2012-04-26 18:57:07Z - Taken code(both fix and tests) written by Antoine and put it in a patch. Done nothing new.  
* [Twisted_2198.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c07/c076cfc485c8ba2f77ebc74a81dcd94bc3eb540d/d43f02b285dcf8cf5a5acec01f4663dea7896a59.patch) (4341 bytes) - added by moijes12 on 2012-04-27 17:21:25Z - I hadn't attached the news file. 

<details><summary>Searchable metadata</summary>

```
trac-id__2198 2198
type__enhancement enhancement
reporter__antoine antoine
priority__normal normal
milestone__ 
branch__branches_sip_multiline_2198 branches/sip-multiline-2198
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1161969038000000 1161969038000000
changetime__1338846570000000 1338846570000000
version__None None
owner__exarkun exarkun
cc__thijs cc__moijes12
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph set owner to @washort|
|-|-|

Say something about Sine, maybe.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_2"></a>|@thijstriemstra commented|
|-|-|

We need unit tests in order to include it into the trunk.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="moijes12's avatar" src="https://avatars.githubusercontent.com/u/4566851?s=50" width="50" height="50">](https://github.com/moijes12)<a name="note_4"></a>|@moijes12 commented|
|-|-|

The test case runs though. I did not want to run the entire set of unit tests as it would required me to uninstall twisted from my system first. Please tell me a if you know of another way.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks.  The patch contains some very simple problems, like the misnamed and undocumented test method, and lines that are too long.  Please check for these easy things before submitting a patch for review.

The change to the parsing seems fine to me.  I'll make the coding standard fixes and apply.  Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [34505]) Branching to 'sip-multiline-2198'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [34506]) Apply patch, plus coding standard fixes
refs [#2198](https://github.com/twisted/twisted/issues/2198)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [34507]) Merge sip-multiline-2198

Author: antoine
Reviewer: exarkun
Fixes: [#2198](https://github.com/twisted/twisted/issues/2198)

Add logic for parsing multiline headers in SIP messages.
